packages/vite/src/node/optimizer/index.ts

```diff
@@ -624,23 +581,58 @@ export function runOptimizeDeps(
           }
         }
 
-        const dataPath = path.join(processingCacheDir, '_metadata.json')
-        writeFile(
-          dataPath,
-          stringifyDepsOptimizerMetadata(metadata, depsCacheDir),
-        )
-
         debug(
           `Dependencies bundled in ${(performance.now() - start).toFixed(2)}ms`,
         )
 
-        return createProcessingResult()
+        return {
+          metadata,
+          async commit() {
+            // Write metadata file, delete `deps` folder and rename the `processing` folder to `deps`
+            // Processing is done, we can now replace the depsCacheDir with processingCacheDir
+            // Rewire the file paths from the temporal processing dir to the final deps cache dir
+
+            if (!fs.existsSync(depsCacheDir)) {
+              fs.mkdirSync(depsCacheDir, { recursive: true })
+            }
+
+            const files = []
+
+            // a hint for Node.js
+            // all files in the cache directory should be recognized as ES modules
+            files.push(
+              fsp.writeFile(
+                path.resolve(depsCacheDir, 'package.json'),
+                JSON.stringify({ type: 'module' }),
```

Comment:
There will be duplicated calls of `JSON.stringify`

Reply:
I feel like doing a writeFile instead of checking existence before writing would be a bit faster, though perf doesn't matter a lot I think since these optimized files are only served to the browser.

Reply:
@sun0day changed it to be directly a string so we can have the same formatting as other files in the deps cache. We could use `JSON.stringify({ type: 'module' }, null, 2) + '\n'` too but the direct string feels better here.

@bluwy I added the check because I think the normal flow would be that this file is already there.

Reply:
Ok, I took that back after the latest changes as it was making the code more complex and I don't think it would make a diff